File: TabCommon/Setting/src/main/java/com/nextzy/setting/view/util/Setting.java
package com.nextzy.setting.view.util;

import android.content.Context;
import android.content.SharedPreferences;
import android.preference.PreferenceManager;
import android.text.TextUtils;

import java.util.Set;

/**
 * Created by「 The Khaeng 」on 24 Oct 2017 :)
 */

public final class Setting {

    public static boolean persistString(Context context, String key, String value, boolean shouldPersist) {
        if (shouldPersist) {
            // Shouldn't store null
            if (TextUtils.equals(value, getPersistedString(context, key, null, shouldPersist))) {
                // It's already there, so the same as persisting
                return true;
            }

            SharedPreferences.Editor editor = getSharedPreferences(context).edit();
            editor.putString(key, value);
            editor.apply();
            return true;
        }
        return false;
    }

    public static String getPersistedString(Context context,String key, String defaultValue, boolean shouldPersist) {
        if (!shouldPersist) {
            return defaultValue;
        }
        return getSharedPreferences(context).getString(key, defaultValue);
    }

    public static boolean persistStringSet(Context context,String key, Set<String> values, boolean shouldPersist) {
        if (shouldPersist) {
            // Shouldn't store null
            if (values.equals(getPersistedStringSet(context, key, null, shouldPersist))) {
                // It's already there, so the same as persisting
                return true;
            }

            SharedPreferences.Editor editor = getSharedPreferences(context).edit();
            editor.putStringSet(key, values);
            editor.apply();
            return true;
        }
        return false;
    }

    public static Set<String> getPersistedStringSet(Context context,String key, Set<String> defaultValue, boolean shouldPersist) {
        if (!shouldPersist) {
            return defaultValue;
        }
        return getSharedPreferences(context).getStringSet(key, defaultValue);
    }

    public static boolean persistInt(Context context, String key, int value, boolean shouldPersist) {
        if (shouldPersist) {
            if (value == getPersistedInt(context, key, ~value, shouldPersist)) {
                // It's already there, so the same as persisting
                return true;
            }

            SharedPreferences.Editor editor = getSharedPreferences(context).edit();
            editor.putInt(key, value);
            editor.apply();
            return true;
        }
        return false;
    }

    public static int getPersistedInt(Context context,String key, int defaultValue, boolean shouldPersist) {
        if (!shouldPersist) {
            return defaultValue;
        }
        return getSharedPreferences(context).getInt(key, defaultValue);
    }


    public static boolean persistFloat(Context context,String key, float value, boolean shouldPersist) {
        if (shouldPersist) {
            if (value == getPersistedFloat(context, key, Float.NaN, shouldPersist)) {
                // It's already there, so the same as persisting
                return true;
            }

            SharedPreferences.Editor editor = getSharedPreferences(context).edit();
            editor.putFloat(key, value);
            editor.apply();
            return true;
        }
        return false;
    }

    public static float getPersistedFloat(Context context,String key, float defaultReturnValue, boolean shouldPersist) {
        if (!shouldPersist) {
            return defaultReturnValue;
        }
        return getSharedPreferences(context).getFloat(key, defaultReturnValue);
    }

    public static boolean persistLong(Context context,String key, long value, boolean shouldPersist) {
        if (shouldPersist) {
            if (value == getPersistedLong(context, key, ~value, shouldPersist)) {
                // It's already there, so the same as persisting
                return true;
            }

            SharedPreferences.Editor editor = getSharedPreferences(context).edit();
            editor.putLong(key, value);
            editor.apply();
            return true;
        }
        return false;
    }

    public static long getPersistedLong(Context context,String key, long defaultValue, boolean shouldPersist) {
        if (!shouldPersist) {
            return defaultValue;
        }
        return getSharedPreferences(context).getLong(key, defaultValue);
    }

    public static boolean persistedBoolean(Context context,String key, boolean value, boolean shouldPersist) {
        if (shouldPersist) {

            if (value == getPersistedBoolean(context, key, !value, shouldPersist)) {
                // It's already there, so the same as persisting
                return true;
            }
            SharedPreferences.Editor editor = getSharedPreferences(context).edit();
            editor.putBoolean(key, value);
            editor.apply();
            return true;
        }
        return false;
    }


    public static boolean getPersistedBoolean(Context context,String key, boolean defaultValue, boolean shouldPersist) {
        if (!shouldPersist) {
            return defaultValue;
        }
        return getSharedPreferences(context).getBoolean(key, defaultValue);
    }


    public static SharedPreferences getSharedPreferences(Context context) {
        return PreferenceManager.getDefaultSharedPreferences(context);
    }


    
}

File: TabCommon/Customize/src/main/java/com/nextzy/tabcustomize/base/extension/Setting.kt
package com.nextzy.tabcustomize.base.extension

import android.content.Context
import android.content.SharedPreferences
import android.preference.PreferenceManager
import android.support.v4.app.Fragment
import com.nextzy.setting.view.util.Setting

/**
 * Created by「 The Khaeng 」on 24 Oct 2017 :)
 */


fun Context.persistString(key: String, value: String, shouldPersist: Boolean = false): Boolean = Setting.persistString(this, key, value, shouldPersist)
fun Fragment.persistString(key: String, value: String, shouldPersist: Boolean = false): Boolean = Setting.persistString(this.context, key, value, shouldPersist)


fun Context.getPersistedString(key: String, defaultValue: String, shouldPersist: Boolean = false): String = Setting.getPersistedString(this, key, defaultValue, shouldPersist)
fun Fragment.getPersistedString(key: String, defaultValue: String, shouldPersist: Boolean = false): String = Setting.getPersistedString(this.context, key, defaultValue, shouldPersist)


fun Context.persistStringSet(key: String, values: Set<String>, shouldPersist: Boolean = false): Boolean = Setting.persistStringSet(this, key, values, shouldPersist)
fun Fragment.persistStringSet(key: String, values: Set<String>, shouldPersist: Boolean = false): Boolean = Setting.persistStringSet(this.context, key, values, shouldPersist)


fun Context.getPersistedStringSet(key: String, defaultValue: Set<String>, shouldPersist: Boolean = false): Set<String> = Setting.getPersistedStringSet(this, key, defaultValue, shouldPersist)
fun Fragment.getPersistedStringSet(key: String, defaultValue: Set<String>, shouldPersist: Boolean = false): Set<String> = Setting.getPersistedStringSet(this.context, key, defaultValue, shouldPersist)


fun Context.persistInt(key: String, value: Int, shouldPersist: Boolean = false): Boolean = Setting.persistInt(this, key, value, shouldPersist)
fun Fragment.persistInt(key: String, value: Int, shouldPersist: Boolean = false): Boolean = Setting.persistInt(this.context, key, value, shouldPersist)


fun Context.getPersistedInt(key: String, defaultValue: Int, shouldPersist: Boolean = false): Int = Setting.getPersistedInt(this, key, defaultValue, shouldPersist)
fun Fragment.getPersistedInt(key: String, defaultValue: Int, shouldPersist: Boolean = false): Int = Setting.getPersistedInt(this.context, key, defaultValue, shouldPersist)


fun Context.persistFloat(key: String, value: Float, shouldPersist: Boolean = false): Boolean = Setting.persistFloat(this, key, value, shouldPersist)
fun Fragment.persistFloat(key: String, value: Float, shouldPersist: Boolean = false): Boolean = Setting.persistFloat(this.context, key, value, shouldPersist)


fun Context.getPersistedFloat(key: String, defaultReturnValue: Float, shouldPersist: Boolean = false): Float = Setting.getPersistedFloat(this, key, defaultReturnValue, shouldPersist)
fun Fragment.getPersistedFloat(key: String, defaultReturnValue: Float, shouldPersist: Boolean = false): Float = Setting.getPersistedFloat(this.context, key, defaultReturnValue, shouldPersist)


fun Context.persistLong(key: String, value: Long, shouldPersist: Boolean = false): Boolean = Setting.persistLong(this, key, value, shouldPersist)
fun Fragment.persistLong(key: String, value: Long, shouldPersist: Boolean = false): Boolean = Setting.persistLong(this.context, key, value, shouldPersist)


fun Context.getPersistedLong(key: String, defaultValue: Long, shouldPersist: Boolean = false): Long = Setting.getPersistedLong(this, key, defaultValue, shouldPersist)
fun Fragment.getPersistedLong(key: String, defaultValue: Long, shouldPersist: Boolean = false): Long = Setting.getPersistedLong(this.context, key, defaultValue, shouldPersist)


fun Context.persistedBoolean(key: String, value: Boolean, shouldPersist: Boolean = false): Boolean = Setting.persistedBoolean(this, key, value, shouldPersist)
fun Fragment.persistedBoolean(key: String, value: Boolean, shouldPersist: Boolean = false): Boolean = Setting.persistedBoolean(this.context, key, value, shouldPersist)


fun Context.getPersistedBoolean(key: String, defaultValue: Boolean, shouldPersist: Boolean = false): Boolean = Setting.getPersistedBoolean(this, key, defaultValue, shouldPersist)
fun Fragment.getPersistedBoolean(key: String, defaultValue: Boolean, shouldPersist: Boolean = false): Boolean = Setting.getPersistedBoolean(this.context, key, defaultValue, shouldPersist)


fun Context.getSharedPreferences(): SharedPreferences? = PreferenceManager.getDefaultSharedPreferences(this)
fun Fragment.getSharedPreferences(): SharedPreferences? = PreferenceManager.getDefaultSharedPreferences(this.context)
